### PR TITLE
Add scripts that publish repos

### PR DESCRIPTION
**Pull request by caesarxuchao**
This is addressing the first step proposed in https://github.com/kubernetes/test-infra/pull/1784#pullrequestreview-22134220, "we get the shell scripts set up and runable by a human."

The next step (in another pull) is letting the publish robot run these scripts. In this pull, I made client-go's publish script to share the code with apimachinery's script. This will simplify the publish robot's code a lot.

Regarding updating depdencies, the publish script will update the client-go's Godeps.json with the just published commit of k8s.io/apimachinery, and will make sure that client-go works with the just published apimachinery. vendor/k8s.io/* and vendor/../glog are removed, to let `go get` users get a single copy of these dependencies in their GOPATH.

See the generated changes at:
https://github.com/kubernetes/apimachinery/pull/8/commits
https://github.com/kubernetes/client-go/pull/133

Side note:
Depends on https://github.com/kubernetes/kubernetes/pull/42084. The publish_client_go.sh will fail if _vendor still exist in staging/client-go. We also need to sync client-go after #42084 is merged and manually add a starting [filter-branch-sha](https://github.com/caesarxuchao/client-go/blob/master/filter-branch-sha).

### CODE REVIEW

**Comment by caesarxuchao** [comment]
HEAD of client-go doesn't build with HEAD of apimachinery, sigh..
```
panic: Double registration of different types for autoscaling/v2alpha1, Kind=ListOptions: old=k8s.io/client-go/pkg/api/v1.ListOptions, new=k8s.io/apimachinery/pkg/apis/meta/v1.ListOptions

goroutine 1 [running]:
panic(0x10cf9c0, 0xc42043d240)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/apimachinery/pkg/runtime.(*Scheme).AddKnownTypeWithName(0xc4203c04c0, 0x128e781, 0xb, 0x128bf80, 0x8, 0x104905f, 0xb, 0x1905be0, 0xc4202e27e0)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/apimachinery/pkg/runtime/scheme.go:194 +0x8e1
k8s.io/apimachinery/pkg/runtime.(*Scheme).AddKnownTypes(0xc4203c04c0, 0x128e781, 0xb, 0x128bf80, 0x8, 0xc4205abc18, 0x4, 0x4)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/apimachinery/pkg/runtime/scheme.go:172 +0x1fb
k8s.io/apimachinery/pkg/apis/meta/v1.AddToGroupVersion(0xc4203c04c0, 0x128e781, 0xb, 0x128bf80, 0x8)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/apimachinery/pkg/apis/meta/v1/register.go:51 +0x46f
k8s.io/client-go/pkg/apis/autoscaling/v2alpha1.addKnownTypes(0xc4203c04c0, 0x0, 0x0)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/pkg/apis/autoscaling/v2alpha1/register.go:47 +0x2e7
k8s.io/apimachinery/pkg/runtime.(*SchemeBuilder).AddToScheme(0x1964f50, 0xc4203c04c0, 0x0, 0x0)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/apimachinery/pkg/runtime/scheme_builder.go:29 +0x58
k8s.io/apimachinery/pkg/runtime.(*SchemeBuilder).AddToScheme-fm(0xc4203c04c0, 0x0, 0x0)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/pkg/api/register.go:72 +0x34
k8s.io/client-go/kubernetes/scheme.AddToScheme(0xc4203c04c0)
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/kubernetes/scheme/register.go:73 +0xc9
k8s.io/client-go/kubernetes/scheme.init.1()
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/kubernetes/scheme/register.go:48 +0x6d
k8s.io/client-go/kubernetes/scheme.init()
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/kubernetes/scheme/register.go:84 +0x181
k8s.io/client-go/kubernetes/typed/apps/v1beta1.init()
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/kubernetes/typed/apps/v1beta1/statefulset.go:173 +0x33
k8s.io/client-go/kubernetes.init()
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/kubernetes/import_known_versions.go:42 +0x3c
k8s.io/client-go/examples/third-party-resources.init()
	/usr/local/google/home/xuchao/go-workspace/src/k8s.io/client-go/examples/third-party-resources/types_test.go:29 +0x51
main.init()
	k8s.io/client-go/examples/third-party-resources/_test/_testmain.go:56 +0x3d
FAIL	k8s.io/client-go/examples/third-party-resources	0.240s
```

**Comment by caesarxuchao** [comment]
cc @sttts @lavalamp you can take a look at the idea, i'm still working on this pull though.

File: mungegithub/mungers/publish_scripts/publish_driver.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_driver.sh]: This shows how i used the other scripts to generate the changes. This script will be removed before merging.

**Comment by caesarxuchao** [comment]
@sttts this is working ^^ Could you give it a pass? Thanks.

File: mungegithub/mungers/publish_scripts/publish_apimachinery.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+fi
+
+TOKEN="${1}"
+# src branch of k8s.io/kubernetes
+SRC_BRANCH="${2:-master}"
+# dst branch of k8s.io/apimachinery
+DST_BRANCH="${3:-master}"
+readonly TOKEN SRC_BRANCH DST_BRANCH
+
+dir=$(dirname "${BASH_SOURCE}")
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_apimachinery.sh]: s/dir/SCRIPT_DIR/

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+fi
+
+TOKEN="${1}"
+# src branch of k8s.io/kubernetes
+SRC_BRANCH="${2:-master}"
+# dst branch of k8s.io/client-go
+DST_BRANCH="${3:-master}"
+readonly TOKEN SRC_BRANCH DST_BRANCH
+
+dir=$(dirname "${BASH_SOURCE}")
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: as above

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script publishes the latest changes in the ${src_branch} of
+# k8s.io/kubernetes/staging/src/client-go to the ${dst_branch} of
+# k8s.io/client-go.
+#
+# The script assumes that the working directory is
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: ... and that k8s.io/apimachinery has been updated

File: mungegithub/mungers/publish_scripts/publish_apimachinery.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+set_github_token "${TOKEN}"
+trap cleanup_github_token EXIT SIGINT
+
+sync_repo "staging/src/k8s.io/apimachinery" "${SRC_BRANCH}"
+# restore the vendor/ folder. k8s.io/* and github.com/golang/glog will be
+# removed from the vendor folder
+restore_vendor
+
+# publish
+git push origin $(git rev-parse --abbrev-ref HEAD)
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_apimachinery.sh]: this push should happen in the driver script at the very end when we are sure that all preparation in **all** scripts was successful.

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    local dir=$(mktemp -d "${TMPDIR:-/tmp/}$(basename $0).XXXXXXXXXXXX")
+
+    new_rev=$(cd ../apimachinery; git rev-parse HEAD)
+
+    while read path rev; do
+        if [[ "${path}" == "k8s.io/apimachinery"* ]]; then
+            echo "${rev}" > "${dir}/rev"
+        fi
+    done < <(jq '.Deps|.[]|.ImportPath + " " + .Rev' -r < "${godeps_json}")
+    echo "$(uniq "${dir}/rev")" > "${dir}/rev"
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: does this work? Better write to a different file first and output to rev here.

Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: Is this file needed at all? can't we use a variable?

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+# be updated with the latest commit hash.
+update_godeps_json
+# restore the vendor/ folder. k8s.io/* and github.com/golang/glog will be
+# removed from the vendor folder
+restore_vendor
+# Smoke test client-go
+basic_tests
+
+# publish
+git push origin $(git rev-parse --abbrev-ref HEAD)
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: as above, at the very end

File: mungegithub/mungers/publish_scripts/publish_driver.sh (hunk 1 above)
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_driver.sh]: Oh, I thought this is called. We definitely need a split between prepare_repos and push_repos.

File: mungegithub/mungers/publish_scripts/util.sh [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+# limitations under the License.
+
+# This file includes functions shared by the each repository's publish scripts.
+
+# sync_repo() cherry picks the latest changes in k8s.io/kubernetes/${filter} to the
+# local copy of the repository to be published.
+#
+# overall flow
+# 1. fetch the current level of k8s.io/kubernetes
+# 2. check out the k8s.io/kubernetes HEAD into a separate branch
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: what does this sentence mean?

File: mungegithub/mungers/publish_scripts/util.sh [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+# 3. rewrite the history on that branch to *only* include staging/src/k8s.io/apimachinery
+# 4. locate all commits between the last time we sync'ed and now
+# 5. switch back to the starting branch
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: starting branch = separate branch from 2. ?

File: mungegithub/mungers/publish_scripts/util.sh [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+# 6. for each commit, cherry-pick it (which will keep authorship) into current branch
+# 7. update metadata files indicating which commits we've sync'ed to
+#
+# The function assumes to be called at the root of the repository that's going to be published.
+sync_repo() {
+    # filter could be staging/src/k8s.io/apimachinery
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: it's a path? Better: a pattern compatible with `git filter-branch --subdirectory-filter`

File: mungegithub/mungers/publish_scripts/util.sh [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    filter="${1}"
+    src_branch="${2}"
+    readonly filter src_branch
+    set -e
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: why this set?

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+source "${dir}"/util.sh
+
+set_github_token "${TOKEN}"
+trap cleanup_github_token EXIT SIGINT
+
+# this currently only updates commit hash of k8s.io/apimachinery
+update_godeps_json() {
+    godeps_json="./Godeps/Godeps.json"
+    local dir=$(mktemp -d "${TMPDIR:-/tmp/}$(basename $0).XXXXXXXXXXXX")
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: with rev being a variable, no need for this anymore

File: mungegithub/mungers/publish_scripts/util.sh [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    dir=$(mktemp -d "${TMPDIR:-/tmp/}$(basename $0).XXXXXXXXXXXX")
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: Wouldn't a variable for `commits` do?

File: mungegithub/mungers/publish_scripts/util.sh [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    git branch -D kube-sync || true
+    git checkout upstream-kube/"${src_branch}" -b kube-sync
+    git reset --hard upstream-kube/"${src_branch}"
+    newKubeSHA=$(git log --oneline --format='%H' kube-sync -1)
+    
+    # this command rewrite git history to *only* include $filter 
+    git filter-branch -f --subdirectory-filter "${filter}" HEAD
+    
+    newBranchSHA=$(git log --oneline --format='%H' kube-sync -1)
+    #git log --no-merges --format='%H' --reverse ${previousBranchSHA}..HEAD^ > ${dir}/commits
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: this line will go I guess

File: mungegithub/mungers/publish_scripts/util.sh [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    rm -rf ./vendor/github.com/golang/glog
+    # this ensures users who get the repository via `go get` won't end up with
+    # multiple copies of k8s.io/ repos. The only copy will be the one in the
+    # GOPATH.
+    # Godeps.json has a complete, up-to-date list of dependencies, so
+    # Godeps.json will be the ground truth for users using godep/glide/dep.
+    rm -rf ./vendor/k8s.io  
+    git add --all
+    # check if there are new contents 
+    if git diff --cached --exit-code &>/dev/null; then
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: woulnd't drop the output

Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: oh, this is not an error case. disregard then. we can drop the output.

File: mungegithub/mungers/publish_scripts/publish_apimachinery.sh (hunk 2 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_apimachinery.sh]: Done.

File: mungegithub/mungers/publish_scripts/publish_apimachinery.sh (hunk 5 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_apimachinery.sh]: Yeah, you are right. I'll write the publish part when integrating with the go code.

File: mungegithub/mungers/publish_scripts/util.sh (hunk 10 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: The man page says $filter has to be a directory. Changed to:
```bash
    # subdirectory in k8s.io/kubernetes, e.g., staging/src/k8s.io/apimachinery
    subdirectory="${1}"
```

File: mungegithub/mungers/publish_scripts/util.sh (hunk 13 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Done.

File: mungegithub/mungers/publish_scripts/util.sh (hunk 14 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Eh, sorry.

File: mungegithub/mungers/publish_scripts/publish_client_go.sh (hunk 4 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: Done.

File: mungegithub/mungers/publish_scripts/publish_client_go.sh (hunk 6 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: Done.

File: mungegithub/mungers/publish_scripts/util.sh (hunk 8 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Changed to:
```
# prerequisites
# 1. we are in the root of the repository to be published
# 2. we are on the branch to be published (let's call it "target-branch")
# overall flow
# 1. fetch the current level of k8s.io/kubernetes
# 2. check out the $src_branch of k8s.io/kubernetes as branch kube-sync
# 3. rewrite the history of branch kube-sync to *only* include code in $subdirectory
# 4. locate all commits between the last time we sync'ed and now
# 5. switch back to the "target-branch"
# 6. for each commit, cherry-pick it (which will keep authorship) into "target-branch"
# 7. update metadata files indicating which commits we've sync'ed to
```

File: mungegithub/mungers/publish_scripts/util.sh (hunk 11 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: removed, mistakenly copied from deads2k's original script.

**Comment by caesarxuchao** [comment]
Comments addressed, @sttts ptal. Thanks!

File: mungegithub/mungers/publish_scripts/util.sh [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    previousBranchSHA=$(cat filter-branch-sha)
+    
+    git remote add upstream-kube https://github.com/kubernetes/kubernetes.git || true
+    git fetch upstream-kube
+    git branch -D kube-sync || true
+    git checkout upstream-kube/"${src_branch}" -b kube-sync
+    git reset --hard upstream-kube/"${src_branch}"
+    newKubeSHA=$(git log --oneline --format='%H' kube-sync -1)
+    
+    # this command rewrite git history to *only* include $subdirectory
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: nit: rewrites

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+git checkout "${DST_BRANCH}"
+
+source "${SCRIPT_DIR}"/util.sh
+
+set_github_token "${TOKEN}"
+trap cleanup_github_token EXIT SIGINT
+
+# this currently only updates commit hash of k8s.io/apimachinery
+update_godeps_json() {
+    godeps_json="./Godeps/Godeps.json"
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: nit: in all files: put `local` in front of function variables.

File: mungegithub/mungers/publish_scripts/publish_client_go.sh [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+    while read path rev; do
+        if [[ "${path}" == "k8s.io/apimachinery"* ]]; then
+            old_revs+="${rev}"$'\n'
+        fi
+    done < <(jq '.Deps|.[]|.ImportPath + " " + .Rev' -r < "${godeps_json}")
+    old_revs=$(echo "${old_revs%%$'\n'}" | sort | uniq)
+    while read old_rev; do
+        sed -i "s|${old_rev}|${new_rev}|g" "${godeps_json}"
+    done <<< "${old_revs}"
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: There is certainly an elegant jq command replacing these 9 lines. I wish I was a better jq wizard.....

**Comment by sttts** [comment]
Up to some nits, it looks good now

File: mungegithub/mungers/publish_scripts/publish_client_go.sh (hunk 18 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/publish_client_go.sh]: Added a TODO

**Comment by caesarxuchao** [comment]
Comments addressed. @lavalamp do you want to take another look?

File: mungegithub/mungers/publish_scripts/util.sh [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+    # Godeps.json has a complete, up-to-date list of dependencies, so
+    # Godeps.json will be the ground truth for users using godep/glide/dep.
+    rm -rf ./vendor/k8s.io  
+    git add --all
+    # check if there are new contents 
+    if git diff --cached --exit-code &>/dev/null; then
+        echo "vendor hasn't changed!"
+        return
+    fi
+    git -c user.name="Kubernetes Publisher" -c user.email="k8s-publish-robot@users.noreply.github.com" commit -m "update vendor/"
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: nit: "Update vendor/"

**Comment by sttts** [comment]
lgtm

File: mungegithub/mungers/publish_scripts/util.sh [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+    
+    git remote add upstream-kube https://github.com/kubernetes/kubernetes.git || true
+    git fetch upstream-kube
+    git branch -D kube-sync || true
+    git checkout upstream-kube/"${src_branch}" -b kube-sync
+    git reset --hard upstream-kube/"${src_branch}"
+    local newKubeSHA=$(git log --oneline --format='%H' kube-sync -1)
+    
+    # this command rewrites git history to *only* include $subdirectory 
+    git filter-branch -f --subdirectory-filter "${subdirectory}" HEAD
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: @caesarxuchao For our OpenShift downstream work we have to be able to cherry-pick commits over multiple repos, i.e. apimachinery, client-go, kube. It would help **a lot** to have the upstream commit hash in the commit msg. This can easily be done like this:

```
git filter-branch -f --msg-filter 'awk 1 - && echo && echo "Kubernetes-commit: ${GIT_COMMIT}"' --subdirectory-filter "${subdirectory}" HEAD
```

Note: `awk 1 -` sanitizes the trailing newline, compare http://backreference.org/2010/05/23/sanitizing-files-with-no-trailing-newline/

Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Why is the "-" in "awk 1 -" significant?

Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Appending the original commit hash changes the new commit hashes, so the metadata in filter-branch-sha becomes wrong. Can we apply this suggestion in a separate PR?

Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: The `-` can be dropped.

Comment (sttts) [inline mungegithub/mungers/publish_scripts/util.sh]: We can move this to another PR, of course.

**Alternative**: get rid of `filter-branch-sha` and use the `Kubernetes-commit: ${previousKubeSHA}` in the commit msg:
```shell
local previousBranchSHA=$(git log --grep "Kubernetes-commit: ${previousKubeSHA}" --pretty=format:"%h")
```

Comment (caesarxuchao) [inline mungegithub/mungers/publish_scripts/util.sh]: Right! Will do that in a followup PR. Just want to get a fresh start :)

**Comment by krzyzacy** [comment]
/lgtm
based off https://github.com/kubernetes/test-infra/pull/2077#issuecomment-283884305

**Comment by caesarxuchao** [comment]
Thanks @krzyzacy !

**Comment by sttts** [comment]
🚀
